daml/Marketplace/Distribution/Auction/Service.daml

```diff
@@ -0,0 +1,201 @@
+{-# LANGUAGE MultiWayIf #-}
+
+module Marketplace.Distribution.Auction.Service where
+
+import DA.Action (foldlA)
+import DA.Foldable (forA_)
+import DA.List (sortOn, mapAccumL)
+import DA.Finance.Asset (AssetDeposit)
+import DA.Finance.Types (Account, Asset, Id)
+import DA.Finance.Utils (fetchAndArchive)
+import Marketplace.Settlement qualified as Settlement
+import Marketplace.Distribution.Auction.Model qualified as Auction
+import Marketplace.Distribution.Auction.Model (Auction(..))
+import Marketplace.Distribution.Auction.Utils (dutchAuction, splitList, withdrawAllocation, updateBidStatus, generateSettlementInstruction)
+import Marketplace.Distribution.Bid.Service qualified as Bid
+import Marketplace.Distribution.Bid.Model qualified as Bid
+import Marketplace.Trading.AllocationAccountRule qualified as Allocation
+import Marketplace.Trading.AllocationAccountRule (AllocationAccountRule(..))
+
+type S = Service
+
+template Service
+  with
+    operator : Party
+    provider : Party
+    customer : Party
+  where
+    signatory operator, provider, customer
+
+    nonconsuming choice OfferBidder : ContractId Bid.Offer
+      with
+        submitter : Party
+        auctionCid : ContractId Auction.T
+        bidder : Party
+      controller submitter
+      do
+        assertMsg ("Unauthorised Submitter - " <> show submitter) (submitter `elem` [provider, customer])
+        Auction{..} <- fetch auctionCid
+        create Bid.Offer with issuer = customer; customer = bidder; ..
+
+    nonconsuming choice RemoveBidderOffer : ()
+      with
+        submitter : Party
+        bidOfferCid : ContractId Bid.Offer
+      controller submitter
+      do
+        assertMsg ("Unauthorised Submitter - " <> show submitter) (submitter `elem` [provider, customer])
+        archive bidOfferCid
+
+    nonconsuming choice RemoveBidder : ()
+      with
+        submitter : Party
+        bidServiceCid : ContractId Bid.S
+      controller submitter
+      do
+        assertMsg ("Unauthorised Submitter - " <> show submitter) (submitter `elem` [provider, customer])
+        archive bidServiceCid
+
+    nonconsuming choice AddLeadBidder : ContractId LeadBidder
```

Comment:
Suggest to rename that to `PublishBid` and you pass in the list of observers it should have.